### ISSUE TITLE
Feature request: streamline request construction

### ISSUE THREAD

**Issue by blixt** [Alamofire/Alamofire#729]
**TL;DR**: Writing code that creates an `Alamofire.Request` from a method, a path, a set of query string parameters, form fields, and (possibly) file uploads is more difficult than expected.

This is a pretty big and undefined issue so I apologize for that in advance, but I'm hoping it might start a constructive discussion on a problem I've run into (or at the very least tell me I'm doing things wrong and there's a better way :)

I'm writing code to interface with the API for our app. To give some context, a request to our API may be `GET`, `POST`, or `DELETE`, it may take query string parameters, form parameters (either as HTTP body query string or multipart fields), and file uploads (as named multipart fields).

I decided to use Alamofire to perform network requests, and early on I found `URLRequestConvertible` to be a useful abstraction. But the moment I needed to upload files my code started branching and I had to separate code out of the `URLRequestConvertible` implementation into separate methods which need to be called when constructing multipart form data.

Here's the function that takes an intent (an `enum`) and converts it to the necessary HTTP information which is used to perform the request with Alamofire:

```swift
func request(intent: Intent, callback: Callback) {
    let requestInfo = self.getRequestInfo(intent)

    // Branch depending on whether there is a file to upload.
    if requestInfo.files.count > 0 {
        Alamofire.upload(
            requestInfo,
            multipartFormData: {
                requestInfo.applyMultipartFormData($0)
            },
            encodingCompletion: { result in
                switch result {
                case let .Success(request, _, _):
                    self.handleResponse(intent, request, callback)
                case let .Failure(error):
                    callback(object: nil, error: error)
                }
            }
        )
    } else {
        let request = Alamofire.request(requestInfo)
        self.handleResponse(intent, request, callback)
    }
}
```

The apparent problem here is that I'm having to branch code outside of my `RequestInfo` class which implements the `URLRequestConvertible` protocol (which initially seemed a promising way to only have to do `Alamofire.request(requestInfo).response…`).

Furthermore, since I have to treat form fields differently between multipart form data and regular URL encoded data in the HTTP body, I'm also branching inside the `URLRequest` property of my `RequestInfo` class:

```swift
var URLRequest: NSURLRequest {
    var request = NSMutableURLRequest(URL: self.url)

    if let token = self.session?.accessToken {
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }

    // Add the query string parameters to the path. Form fields or files are not added here.
    if self.queryString.count > 0 {
        (request, _) = Alamofire.ParameterEncoding.URL.encode(request, parameters: self.queryString)
    }

    // Apply method after query string so that Alamofire doesn't put it in the HTTP body.
    request.HTTPMethod = self.method.rawValue

    // Apply form data here if there are no files to upload.
    if self.form.count > 0 && self.files.count == 0 {
        precondition(self.method == .POST, "Form data can only be added to POST requests")
        (request, _) = Alamofire.ParameterEncoding.URL.encode(request, parameters: self.form)
    }

    return request
}
```

As you can see I'm only conditionally including the form data here (otherwise I apply it in the `applyMultipartFormData` method). This makes the code a bit smelly on my end. Also, it would be nice to have a better way to control what goes on the query string and what goes in the form data, even when making a regular POST request, other than the workaround I'm using (or reimplementing a custom encoder).

So my feature request is essentially a streamlined way to construct `Alamofire.Request` which is either a file upload or a regular GET/POST request without having to branch the code so much. I hope this input is relevant and let me know if there's any additional information I can provide, or if there's any way I can help out.

**Comment by cnoon** [comment]
Hi @blixt,

Thank you for taking the time to put together this really well thought out issue. You have some very good feedback and use cases that others could learn from about the way Alamofire is intended to be used and is designed.

### Issue

I think the biggest issue you are fighting at the moment is that you are trying to funnel three very different types of requests through the same root method. Additionally, you're trying to build all three requests using the same `URLRequestConvertible` implementation. Now I completely get why you are trying to boil your API calls down, but you've passed the point of no return with reusability IMO.

### Splitting Up URLRequestConvertible

Why not create three different types of objects with their own implementations rather than trying to jam them all into one? You have three different types of requests to make, so have three different wrappers for these types of requests? This would certainly solve the issues you are fighting against at the moment. I think the code samples in the [API Parameter Abstraction](https://github.com/Alamofire/Alamofire/tree/swift-2.0#api-parameter-abstraction) and the [CRUD](https://github.com/Alamofire/Alamofire/tree/swift-2.0#crud--authorization) examples demonstrate how to sensibly encapsulate similar chunks of logic. What I would not encourage though is to use a single enumeration to work for EVERY single API. It gets too daunting to handle all the cases, especially when your request types are so drastically different.

### Don't Funnel into a Single Alamofire Call

Another suggestion I have is to not try to wrap your entire API into a single call into Alamofire. Instead, I tend to wrap APIs as follows:

```swift
class SomeService {
    static func callFirstService(parameters: [String: AnyObject], completion: ModelObject? -> Void) {
         // Implement some Alamofire.request() call fully tailored to this service call
         // Attempt to parse the logic into the model object and serve it back to the callback
    }

    static func callSecondService(parameters: [String: AnyObject], completion: ModelObject? -> Void) {
         // Implement some Alamofire.request() call fully tailored to this service call
         // Attempt to parse the logic into the model object and serve it back to the callback
    }

    // So on and so forth
}
```

> This approach let's you tailor each `Request` to the service you are calling. Alamofire is MUCH easier to use with this approach rather than funneling these calls into a single call to execute the request.

### Embrace Chaining

The approach you've taken really goes against the idea of method chaining. Alamofire makes it really easy to customize the setup of a request, validation and response handling. It is not structured at all to handle all these cases in a single API. This is one of the most powerful parts of the Alamofire design, and you are really working against it rather than embracing it.

### Summary

I really appreciate all your feedback. It's very helpful to see where people are running into design issues when leveraging Alamofire to call their APIs. I would really encourage you to read through this feedback carefully and consider these design suggestions. Alamofire can easily achieve what you need, but gets complicated with the way your trying to funnel your calls through it.

**Comment by blixt** [comment]
Thank you for your response @cnoon.

I can see why you'd say they are fundamentally different requests, but that is at a technical level which shouldn't need to be so prevalent for the API user. The different requests I mentioned above are not, from a semantical perspective, very different. As an example, let's have a look at Facebook's API endpoint for creating a photo in an album (https://developers.facebook.com/docs/graph-api/reference/v2.4/album/photos#publish).

When creating a photo you get to choose whether you want to set the `source` field (which is multipart form data) or if you set the `url` field (which is just a string parameter). At its simplest I might expect this to look something like:

```swift
let request = Request.post("https://graph.facebook.com/v2.4/\(albumId)/photos")
// pick either:
request.file(myPhotoFile, field: "source").…
// or:
request.form("http://lorempixel.com/400/400/cats/", field: "url").…
```

I would say the above exactly captures the choice given by the Facebook API. Now what actually happens is of course different under the hood (and the technical implementation of multipart data is more complex than simple POST query string data). But as the person making HTTP calls I really don't care, and I don't think this API stops people who want to customize further from doing so (as you said, chaining can be powerful and that applies here too).

I guess my biggest gripe is that Alamofire split out a content encoding difference (query string vs. multipart) into different high level code paths, and I have to do a lot of branching in my code to accommodate for that fact. Now if I'm missing something obvious, please let me know.

Finally, I still think it's strange that Alamofire doesn't let me choose what data to put in the query string and what to put in the HTTP body. Having that ability is quite important for logging transparency reasons.

**Comment by cnoon** [comment]
Hi @blixt,

Now we're narrowing it down. I couldn't agree more with your comments about the Facebook API. That's exactly the message I was trying to relay in my "Don't funnel into a single call" section. We're on the same page there. Alamofire makes that quite simple with the exception of `MultipartFormData`. I think my biggest take away here is your frustration with the two-step handling of `MultipartFormData`. 

It was designed that way because it has to be if you are uploading a large amount of data. The data needs to be encoded to disk before you can create the request. Since the request isn't actually created until the encoding completes, the method call cannot return the request to allow for chaining. The only way it could return a request to allow for chaining would be to encode the data in memory or on disk on the caller thread and directly create the upload request and return it. This has very problematic implications with blocking the caller thread for a significant amount of time. This is why uploading `MultipartFormData` was designed as a two-step process.

Now, you could certainly make this a one step process by pre-encoding the `MultipartFormData` manually, then handling off the already encoded data to the regular upload methods (data, file, stream). This would allow you to chain both approaches in the same manner. If you have other suggestions on how to handle the two-step problem in a more elegant manner, I'm all ears. IMO there isn't a better way to do this unless all NSURLSessionTask creation was deferred until `Request.resume()` was technically called. This however would require a complete re-write of the Alamofire core logic and would have MANY other downsides.

As for your last comment of controlling what data is put into the query string and what is put into the body, I tend to disagree. If you need fine grained control, create your own `NSURLRequest` objects with the `ParameterEncoding` enum and then pass that `NSURLRequest` into the `URLRequestConvertible` methods. The convenience APIs designed to do it all in one step (request, download, upload) are a bit too convenient for you. With that said, there is currently a PR open to add another URL encoding case that will at least allow you to specify that the encoded parameters should go into the URL and not the body regardless of the Method.

Great discussion...thanks for all the feedback. So far I don't have any actionable items here. If you can come up with some suggestions on the `MultipartFormData` design, I'm happy to revisit the design of the upload methods.

**Comment by blixt** [comment]
I think I might not have made my original example very clear then, because I'm not doing anything different in my code than the Facebook example I made. I have one data structure per endpoint (technically it's an `enum`). The problem for me is that I had to do quit a bit of abstraction on my end rather than relying solely on Alamofire. One issue is that `URLRequestConvertible` simply can't represent the full state of a request (for multipart form data specifically, and as a lesser issue the query string vs form data), which requires me to leak information outside that class.

For the case of `MultipartFormData`, I am of the opinion that a library should be designed to make common cases easy, and I believe the streaming-from-disk you speak of doesn't occur until something is 10 MB or larger. This seems to me to be a pretty rare case with the exception of video, and shouldn't have to hamper all other use of the library.

In terms of API design, I would say it sounds like you guys need streamable HTTP body support more than specifically built-in multipart handling. Now I might be unaware of `NSURLSession` internals that stop you from doing this, but it seems to me that for people who need to do something more special in regards to how the body is serialized, they can implement a protocol. Something like this:

```swift
let serializer = MyCustomMultipartFormDataSerializer()
serializer.fields["source"] = myLocalCatVideoFileURL
serializer.fields["title"] = "Cats Are Awesome"
Request.post("http://example.com/upload_video").content(serializer)
```

Obviously the serializer would have to be responsible for providing `Content-Type`, `Content-Length` and other necessary headers for this to work, but the meat of the implementation would be in allowing Alamofire to asynchronously read and send data from the serializer to the HTTP endpoint. This could also evolve into gzip, etc. support, and support chaining (data -> gzip -> multipart). Again, this all comes with the caveat that it's possible to do this with the iOS internals.

Anyway, the reason for this discussion is the snag that I have to branch my code based on whether I include a string or a file. I understand that the underlying technical code path is very different, but libraries are there to hide those details. :) If I want to express the Facebook API calls above, it shouldn't need to be more code than the examples I gave. You mention that `request` and `upload` are too convenient for me – could you give me an example of the less convenient API that achieves the same thing? I'd like to figure out the best way to make use of the Alamofire API in a way that makes my code more succinct.

Finally, I'd be curious to hear your arguments for why you disagree on the lack of control over query string and form data distinction. As a backend developer maybe I see them as more separate than you do, but I can say that query strings and form data are very often separate data stores in most libraries that handle HTTP requests, and you'd be lucky to not run into issues where the backend expects the data to be in one or the other. Just having a flag to control all of the parameters is a step in the right direction, but it makes me think that you might as well just take a second step and instead of making it a flag to pick one or the other, making it two sets of data. The API would be equally expressive in cases where you just want one or the other, but with the additional ability to specify both for cases like mine.

**Comment by cnoon** [comment]
Hi @blixt,

Your serializer suggestions are good, but lack the context around the limitations of this type of implementation on iOS. Subclassing `NSInputStream` is VERY fragile and simply not recommended. We've learned the hard way from our trials with AFNetworking. This is why we went with the approach we did in Alamofire. It is much more robust and reliable. You still aren't addressing the fundamental issue of encoding the data before creating the `Request`. You cannot do it after you create the `Request`. You're ideas are all really good, but they aren't possible to implement. Believe me, I considered them heavily before building the current implementation.

As for your challenges with using `ParameterEncoding` in both the query string and the HTTP body, you can still handle this case by using a combination of `ParameterEncoding` and `URLRequestConvertible`. Alamofire has always been designed to make the common cases as simple as possible, while being flexible enough to still allow you (the power user) to perform any operations that you need by having access to the underlying URL Loading System.

I appreciate all your feedback in this thread. I think it is very useful and well thought out. As a takeaway, I would challenge you with a couple things. 1) Try to work with Alamofire rather than working against it. It's not designed to be perfectly convenient for every use case. It's designed to be convenient for the common cases, yet give you the flexibility to control what you need. Layer the convenience levels around it in your own APIs to abstract the Alamofire complexities. 2) Open a pull request with your ideas and suggestions on how we can make improvements. We are in no means locked into any way of doing things. We take the submitted PRs very seriously and love it when new ideas and designs come in from the community.

Cheers.
